modules/siddhi-core/src/main/java/io/siddhi/core/table/record/AbstractQueryableRecordTable.java

```diff
@@ -431,103 +335,149 @@ public void updateOrAdd(ComplexEventChunk<StateEvent> updateOrAddingEventChunk,
-                            updateConditionParameterMaps, recordTableCompiledUpdateSet.getUpdateSetMap(),
-                            updateSetParameterMaps, addingRecords);
-                } else {
-                    updateOrAdd(recordStoreCompiledCondition.compiledCondition, updateConditionParameterMaps,
-                            recordTableCompiledUpdateSet.getUpdateSetMap(), updateSetParameterMaps, addingRecords);
-                }
+                cacheResults = cacheTable.find(compiledConditionWithCache.getCacheCompileCondition(),
+                        matchingEvent);
+                return cacheResults;
             } finally {
-                readWriteLock.writeLock().unlock();
-            }
-            if (cachedTable.size() > maxCacheSize) {
-                isCacheEnabled = false;
-                log.warn(siddhiAppContext.getName() + ": " + cacheTableDefinition.getId() + " size is now " +
-                        cachedTable.size() + " which is bigger than cache table size defined as " + maxCacheSize +
-                        ". So cache is now disabled");
+                readWriteLock.readLock().unlock();
             }
         } else {
-            if (recordTableHandler != null) {
-                recordTableHandler.updateOrAdd(timestamp, recordStoreCompiledCondition.compiledCondition,
-                        updateConditionParameterMaps, recordTableCompiledUpdateSet.getUpdateSetMap(),
-                        updateSetParameterMaps, addingRecords);
-            } else {
-                updateOrAdd(recordStoreCompiledCondition.compiledCondition, updateConditionParameterMaps,
-                        recordTableCompiledUpdateSet.getUpdateSetMap(), updateSetParameterMaps, addingRecords);
+            if (log.isDebugEnabled()) {
+                log.debug(siddhiAppContext.getName() + ": store table size is bigger than cache.");
             }
-        }
+            if (cacheEnabled && compiledConditionWithCache.isRouteToCache()) {
+                if (log.isDebugEnabled()) {
+                    log.debug(siddhiAppContext.getName() + ": cache constraints satisfied. Checking cache");
+                }
+                readWriteLock.readLock().lock();
+                try {
+                    cacheResults = cacheTable.find(compiledConditionWithCache.getCacheCompileCondition(),
+                            matchingEvent);
+                    if (cacheResults != null) {
+                        if (log.isDebugEnabled()) {
+                            log.debug(siddhiAppContext.getName() + ": cache hit. Sending results from cache");
+                        }
+                        return cacheResults;
+                    }
+                } finally {
+                    readWriteLock.readLock().unlock();
+                }
+                // cache miss
+                if (log.isDebugEnabled()) {
+                    log.debug(siddhiAppContext.getName() + ": cache miss. Loading from store");
```

Comment:
Also, use the query name in the log if there are queries that work with the tabes then we will not be able to differentiate.

Reply:
How to get the query name inside find func?